pkg/runtime/controller.go

```diff
@@ -233,12 +248,58 @@ func (c *Controller) validHandlerConfigs() map[string]*cpb.Handler {
 			Params:  cfg,
 		}
 	}
+	if glog.V(3) {
+		glog.Infof("handler = %v", handlerConfig)
+	}
 	return handlerConfig
 }
 
+// maxAttributeManifests is the likely maximum number of manifests in the config.
+// It is used to avoid reallocation on config change.
+const maxAttributeManifests = 5
+
+// processAttributeManifests loads attribute manifests to produce an AttributeDescriptorFinder.
+// TODO in apply events mark if AttributeManifestKind changes, so this step is skipped.
+// attribute manifests are not expected to change often.
+func (c *Controller) processAttributeManifests() expr.AttributeDescriptorFinder {
+	attrs := make([]*cpb.AttributeManifest, 0, maxAttributeManifests)
+	// check rules and ensure only good handlers and instances are used.
+	// record handler - instance associations
+	nattributes := 0
+	for k, cfg := range c.configState {
+		if k.Kind != AttributeManifestKind {
+			continue
+		}
+		attrs = append(attrs, cfg.(*cpb.AttributeManifest))
```

Comment:
shall we just flatten all attributes in the AttributeManifest into a map ? GetAttribute might get faster

Reply:
yes, was it was waiting on smarter  config change for attributes, did both.